### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/RootSystem/Defs): Properties of pairs of roots

### PR DESCRIPTION
**Pull request by ScottCarnahan**
Given a pair of roots, we can describe the "angle" between them using the pairing between root `i` and coroot `j` together with the reverse pairing.  This PR adds language for describing the various cases.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by ocfnash — COMMENTED** [review]
In addition to the nitpicks highlighted below, I think it would be worth using `RootPairing.pairing` in the statement of `RootPairing.eq_of_forall_coroot_root_eq`.

**Comment by ScottCarnahan** [comment]
I tried putting `pairing` into the statement of `RootPairing.eq_of_forall_coroot_root_eq`, but fixing the `rw` seems to require a some wrestling.

**Comment by ocfnash** [comment]
> I tried putting `pairing` into the statement of `RootPairing.eq_of_forall_coroot_root_eq`, but fixing the `rw` seems to require some wrestling.

I just pushed [a commit](https://github.com/leanprover-community/mathlib4/pull/10666/commits/43c5ec0486ceeb0a99ddabb124a2814bc4e971e2) which does this wrestling (we can't really use a definition until we write its API).

**Review by ocfnash — APPROVED** [review]
I am happy with these changes. However as I have pushed some commits I think we need another reviewer's opinion before sending them to bors.

File: Mathlib/LinearAlgebra/RootSystem/Basic.lean [hunk 1]
```diff
@@ -113,23 +129,25 @@ lemma reflection_dualMap_eq_coreflection :
   ext n m
   simp [coreflection_apply, reflection_apply, mul_comm (P.toLin m (P.coroot i))]
 
-variable [Finite ι]
+lemma isCrystallographic_iff :
+    P.IsCrystallographic ↔ ∀ i j, ∃ (z : ℤ), z = P.pairing i j := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/RootSystem/Basic.lean:133]:
```suggestion
    P.IsCrystallographic ↔ ∀ i j, ∃ z : ℤ, z = P.pairing i j := by
```
(wasn't there a discussion about `∃ x, x = ...` vs `∃ x, ... = x` recently? I can't find it now, but maybe check if this matches the conclusion, if you can find it)

File: Mathlib/LinearAlgebra/RootSystem/Basic.lean [hunk 2]
```diff
@@ -113,23 +129,25 @@ lemma reflection_dualMap_eq_coreflection :
...
+  · simpa [AddSubgroup.mem_zmultiples_iff] using h i (mem_range_self j)
+  · simpa [← hj, AddSubgroup.mem_zmultiples_iff] using h i j
 
-/-- Even though the roots may not span, coroots are distinguished by their pairing with the
-roots. The proof depends crucially on the fact that there are finitely-many roots.
+variable [Finite ι]
 
-Modulo trivial generalisations, this statement is exactly Lemma 1.1.4 on page 87 of SGA 3 XXI.
-See also `RootPairing.injOn_dualMap_subtype_span_root_coroot` for a more useful restatement. -/
-lemma eq_of_forall_coroot_root_eq [NoZeroSMulDivisors ℤ M] (i j : ι)
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/RootSystem/Basic.lean:123]: Do we want to deprecate `eq_of_forall_coroot_root_eq` before removing it?

File: Mathlib/LinearAlgebra/RootSystem/Defs.lean [hunk 3]
```diff
@@ -130,3 +130,47 @@ def reflection : M ≃ₗ[R] M :=
...
+
+/-- A pair of roots is definite if there is a unique (up to scalars) definite reflection-invariant
+form on their span. This happens precisely when the Coxeter weight is strictly between `0` and `4`.
+-/
+def IsDefinite : Prop := 0 < coxeterWeight P i j ∧ coxeterWeight P i j < 4
+
+/-- A pair of roots is ultraparallel if its Coxeter weight is strictly greater than 4.
+Symmetrization induces a nondegenerate indefinite form on the span, and reflections generate an
+infinite dihedral group. -/
+def IsUltraParallel : Prop := 4 < coxeterWeight P i j
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/RootSystem/Defs.lean:174]: That's a lot of definitions without API.

Comment (ScottCarnahan) [reply Mathlib/LinearAlgebra/RootSystem/Defs.lean:174]: Should I cut some of it out for a later PR?

Comment (jcommelin) [reply Mathlib/LinearAlgebra/RootSystem/Defs.lean:174]: @ScottCarnahan maybe you already have more material? In that case, would you mind opening another PR that has all the material? That helps with reviewing.

For the current PR, I would suggest keeping the first 3 defs `pairing`, `coxeterWeight`, and `IsOrthogonal`, and writing a few API lemmas about them like `coxeterWeight_swap` and `IsOrthogonal_comm` and `IsOrthogonal.symm`.

I guess the other definitions will come with more serious API.

**Review by jcommelin — APPROVED** [review]
LGTM, one more comment.

Thanks a lot! Excited about where Lie theory in mathlib is going!

bors d+

**Pull request closed without merging**